tempodb/backend/gcs/gcs.go

```diff
@@ -186,6 +190,112 @@ func (rw *readerWriter) List(ctx context.Context, keypath backend.KeyPath) ([]st
 	return objects, nil
 }
 
+// ListBlocks implements backend.Reader
+func (rw *readerWriter) ListBlocks(ctx context.Context, keypath backend.KeyPath) (blockIDs []uuid.UUID, compactedBlockIDs []uuid.UUID, err error) {
+	span, ctx := opentracing.StartSpanFromContext(ctx, "readerWriter.ListBlocks")
+	defer span.Finish()
+
+	keypath = backend.KeyPathWithPrefix(keypath, rw.cfg.Prefix)
+	prefix := path.Join(keypath...)
+	if len(prefix) > 0 {
+		prefix += "/"
+	}
+
+	bb := blockboundary.CreateBlockBoundaries(rw.cfg.ListBlocksConcurrency)
+
+	errChan := make(chan error, len(bb))
+	wg := sync.WaitGroup{}
+	mtx := sync.Mutex{}
+
+	var min uuid.UUID
+	var max uuid.UUID
+
+	for i := 0; i < len(bb)-1; i++ {
+		min = uuid.UUID(bb[i])
+		max = uuid.UUID(bb[i+1])
+
+		wg.Add(1)
+		go func(min, max uuid.UUID) {
+			defer wg.Done()
+
+			iter := rw.bucket.Objects(ctx, &storage.Query{
+				Prefix:      prefix,
+				Delimiter:   "",
+				Versions:    false,
+				StartOffset: prefix + min.String(),
```

Comment:
Could we make use of EndOffset?  Without it it seems like there is redundant work.  For 4 shards the first request would list 100% of the bucket, the second request would list 75% of the bucket, etc.  I think you could use EndOffset=max.String(), and then that block=max gets picked up in the next shard where it equals min.  It gets a little tricky for max UUID 0xFFFF... because EndOffset is exclusive so it must not be present on the last request.  I haven't looked into the block boundaries not sure if that is actually a concern.

Reply:
There is currently an early return of we reach a block ID that is out of the scope for the particular routine.  In paged results we would still throw away the last part of a page, but not quite as bad as you mention.

Reply:
Right, I am thinking about overhead on the object storage backend.  The implementation for S3 seems good, the work is split across multiple requests with the continuation token, and it will break after reaching the shard max (up to only 1 page of results thrown away).    But for GCS it looks like a single call, and even though we stop processing the results for the shard, the work has already been done by the object storage to provide them.

Reply:
Ah interesting.  I'd thought the iteration token worked similarly, but you could be right.  I've made the adjustment to include the `EndOffset`.